test-framework/vault/src/main/java/io/quarkus/vault/VaultTestExtension.java

```diff
@@ -0,0 +1,298 @@
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.junit.jupiter.api.extension.ExtensionContext.Namespace.GLOBAL;
+import static org.testcontainers.containers.BindMode.READ_ONLY;
+import static org.testcontainers.containers.PostgreSQLContainer.POSTGRESQL_PORT;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Optional;
+
+import javax.sql.DataSource;
+
+import org.jboss.logging.Logger;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.extension.BeforeAllCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.testcontainers.containers.Container;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.PostgreSQLContainer;
+
+import io.quarkus.vault.runtime.VaultManager;
+import io.quarkus.vault.runtime.client.TestVaultClient;
+import io.quarkus.vault.runtime.client.VaultClientException;
+import io.quarkus.vault.runtime.client.dto.VaultInit;
+import io.quarkus.vault.runtime.client.dto.VaultSealStatus;
+import io.quarkus.vault.runtime.config.VaultServerRuntimeConfig;
+
+public class VaultTestExtension implements BeforeAllCallback, ExtensionContext.Store.CloseableResource {
+
+    private static final Logger log = Logger.getLogger(VaultTestExtension.class.getName());
+
+    private static boolean started = false;
+
+    static final String DB_NAME = "mydb";
+    static final String DB_USERNAME = "postgres";
+    static final String DB_PASSWORD = "bar";
+    static final String SECRET_VALUE = "s\u20accr\u20act";
+    static final String DEFAULT_VAULT_VERSION = "1.2.2";
+    static final int VAULT_PORT = 8200;
+    static final int MAPPED_POSTGRESQL_PORT = 6543;
+    static final String VAULT_AUTH_USERPASS_USER = "bob";
+    static final String VAULT_AUTH_USERPASS_PASSWORD = "sinclair";
+    static final String SECRET_PATH_V1 = "secret";
+    static final String SECRET_PATH_V2 = "secret-v2";
+    static final String VAULT_DBROLE = "mydbrole";
+    static final String APP_SECRET_PATH = "foo";
+    static final String APP_CONFIG_PATH = "config";
+    static final String VAULT_POLICY = "mypolicy";
+    static final String POSTGRESQL_HOST = "mypostgresdb";
+    static final String VAULT_URL = (useTls() ? "https" : "http") + "://localhost:" + VAULT_PORT;
+    static final String SECRET_KEY = "secret";
+
+    public static final String TMP_VAULT_POSTGRES_CREATION_SQL_FILE = "/tmp/vault-postgres-creation.sql";
+    public static final String TMP_VAULT_CONFIG_JSON_FILE = "/tmp/vault-config.json";
+    public static final String TMP_VAULT_POLICY_FILE = "/tmp/vault.policy";
+    public static final String TMP_POSTGRES_INIT_SQL_FILE = "/tmp/postgres-init.sql";
```

Comment:
Hmmm, Vault requires PostgreSQL to work?

Reply:
yes. for the dynamic db credentials use case: you give vault an admin access to your db, and vault will create the user and its password. that is why in that use case, you do not configure any user on your datasource; it is part of the response from the credential provider.